bench_runner/hpt.py

```diff
@@ -0,0 +1,339 @@
+  New Orleans, LA, USA, 2012, pp. 1-12,
+  doi: 10.1109/HPCA.2012.6169043.
+
+This is largely a direct port of the bash implementation available here:
+
+  https://github.com/cirosantilli/parsec-benchmark/tree/master/toolkit/hpt
+
+This approach is a more robust way to measure overall effectiveness across a
+number of benchmarks. It is still biased in that the benchmarks should be a
+representative sample, but it accounts for the fact that some benchmarks are
+more reproducible and reliable than others.
+
+It has been modified so that each benchmark can have a different number of
+samples (the original code assumed the matrix was rectangular, but there is
+nothing about the method itself that should require that).
+
+"""
+
+import io
+import functools
+import json
+import os
+from typing import Dict, Optional, Tuple
+
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+ACC_MAXSU = 2
+
+
+def load_from_json(
+    json_path: os.PathLike,
+) -> Dict[str, NDArray[np.float64]]:
+    with open(json_path) as fd:
+        content = json.load(fd)
+
+    return load_data(content)
+
+
+def load_data(data) -> Dict[str, NDArray[np.float64]]:
+    results = {}
+    for benchmark in data["benchmarks"]:
+        if "metadata" in benchmark:
+            name = benchmark["metadata"]["name"]
+        else:
+            name = data["metadata"]["name"]
+        values = []
+        for run in benchmark["runs"]:
+            values.extend(run.get("values", []))
+        results[name] = np.array(values, dtype=np.float64)
+
+    return results
+
+
+def create_matrices(
+    a: Dict[str, NDArray[np.float64]], b: Dict[str, NDArray[np.float64]]
+) -> Tuple[Dict[str, NDArray[np.float64]], Dict[str, NDArray[np.float64]]]:
+    benchmarks = sorted(list(set(a.keys()) & set(b.keys())))
+    return {bm: a[bm] for bm in benchmarks}, {bm: b[bm] for bm in benchmarks}
+
+
+def qnorm(p: float) -> float:
+    """
+    quantile function of standard norm distribution
+    """
+
+    if p <= 0.0 or p >= 1.0:
+        raise ValueError(f"{p} is out of range 0, 1")
+
+    if p == 0.5:
+        return 0.0
+
+    y = -np.log(4.0 * p * (1.0 - p))
+    b = [
+        1.570796288,
+        0.03706987906,
+        -0.0008364353589,
+        -0.0002250947176,
+        0.000006841218299,
+        0.000005824238515,
+        -0.00000104527497,
+        0.00000008360937017,
+        -0.000000003231081277,
+        0.00000000003657763036,
+        0.0000000000006936233982,
+    ]
+    u = 0.0
+    pow = 1.0
+
+    for b0 in b:
+        pow *= y
+        u += pow * b0
+
+    u = np.sqrt(u)
+    if p < 0.5:
+        u *= -1.0
+
+    return u
+
+
+def cdfnorm(x: float) -> float:
+    """
+    an approximation of cumulative density function for standard norm distribution
+    """
+
+    a1 = 0.31938153
+    a2 = -0.356563728
+    a3 = 1.781477937
+    a4 = -1.821255978
+    a5 = 1.330274429
+    pi = 3.141592653589793238462643
```

Comment:
Could you just use ``math.pi`` here? These constants in general feel a little 'magic', is there a CDF function in the stdlib or numpy that could be used instead?

Reply:
There is a CDF function in Scipy, with different behavior as to how the wraparound is handled. So it's both too big a dependency for this little thing, IMHO, and it doesn't do what we need anyway.

But, yes, could use `math.pi`.